collector/logind_linux.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nologind
+
+package collector
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/godbus/dbus"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	logindSubsystem = "logind"
+	dbusObject      = "org.freedesktop.login1"
+	dbusPath        = "/org/freedesktop/login1"
+)
+
+var (
+	// Taken from logind as of systemd v229
+	attrRemoteValues = []string{"true", "false"}
+	attrTypeValues   = []string{"unspecified", "tty", "x11", "wayland", "mir", "web"}
+	attrClassValues  = []string{"user", "greeter", "lock-screen", "background"}
+)
+
+type logindCollector struct {
+	sessionsDesc *prometheus.Desc
+}
+
+type logindDbus struct {
+	conn   *dbus.Conn
+	object dbus.BusObject
+}
+
+type logindSession struct {
+	seat        string
+	remote      string
+	sessionType string
+	class       string
+}
+
+// Struct elements must be public for the reflection magic of godbus to work
+type logindSessionEntry struct {
+	SessionId         string
+	UserId            uint32
+	UserName          string
+	SeatId            string
+	SessionObjectPath dbus.ObjectPath
+}
+
+type logindSeatEntry struct {
+	SeatId         string
+	SeatObjectPath dbus.ObjectPath
+}
+
+func init() {
+	Factories["logind"] = NewLogindCollector
+}
+
+// Takes a prometheus registry and returns a new Collector exposing
+// logind statistics.
+func NewLogindCollector() (Collector, error) {
+	sessionsDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, logindSubsystem, "sessions"),
+		"Number of sessions registered in logind.", []string{"seat", "remote", "type", "class"}, nil,
+	)
+
+	return &logindCollector{
+		sessionsDesc: sessionsDesc,
+	}, nil
+}
+
+func (lc *logindCollector) Update(ch chan<- prometheus.Metric) error {
+	c, err := newDbus()
+	if err != nil {
+		return fmt.Errorf("unable to connect to dbus: %s", err)
+	}
+	defer c.conn.Close()
+
+	seats, err := listSeats(c)
```

Comment:
On normal systems, there are exactly two seats: The empty seat "", which is assigned to remote sessions, and "seat0", which is used for local sessions (both tty and x11).

Other seats may be configured by the administrator by bundling devices (usually a mouse, a keyboard and a monitor), but such setups aren't very common.

The seat attribute is not really important to me, if you think this might be a problem, I can drop it.